src/expr/src/scalar/func/impls/string.rs

```diff
@@ -921,3 +926,103 @@ sqlfunc!(
         panic!("{}", a)
     }
 );
+
+/// QuoteIdent is transliterated from postgres/src/backend/utils/adt/ruleutils.c
+#[derive(
+    Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect, Arbitrary,
+)]
+pub struct QuoteIdent;
+
+impl LazyUnaryFunc for QuoteIdent {
+    fn eval<'a>(
+        &'a self,
+        datums: &[Datum<'a>],
+        temp_storage: &'a RowArena,
+        a: &'a MirScalarExpr,
+    ) -> Result<Datum<'a>, EvalError> {
+        let d = a.eval(datums, temp_storage)?;
+        if d.is_null() {
+            return Ok(Datum::Null);
+        }
+        let val: &str = d.unwrap_str();
+
+        let mut buf = LexBuf::new(val);
+
+        let mut safe = match buf.next() {
+            Some(ch) if matches!(ch, 'a'..='z' | '_') => true,
+            _ => false,
+        };
+
+        let mut nquotes = 0;
+
+        while let Some(ch) = buf.next() {
+            safe = safe && matches!(ch, 'a'..='z' | '0'..='9' | '_');
+            if ch == '"' {
+                nquotes += 1;
+            }
+        }
+
+        safe = safe && {
+            match mz_sql_parser::keywords::Keyword::from_str(&val) {
+                Err(_) => true,
+                Ok(kw) => !kw.is_sometimes_reserved(),
+            }
+        };
+
+        if safe {
+            return Ok(d);
+        }
+
+        let mut r = String::with_capacity(val.len() + nquotes + 2);
+
+        r.push('"');
+
+        for ch in val.chars() {
+            if ch == '"' {
+                r.push('"');
+            }
+            r.push(ch);
+        }
+
+        r.push('"');
```

Comment:
Wait, can we not use the `Display` implementation of `Ident` in the SQL parser? That _should_ do exactly the same thing you've got going here; if not it's a bug!

Reply:
Oh derp––I have remembered this at other time and forgotten it this morning.